sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts

```diff
@@ -0,0 +1,146 @@
+*/
+
+import {
+  ServiceBusClient,
+  delay,
+  ReceivedMessageWithLock,
+  SessionReceiver,
+  MessagingError
+} from "@azure/service-bus";
+import * as dotenv from "dotenv";
+import { env } from "process";
+import { AbortController, AbortSignalLike } from "@azure/abort-controller";
+
+dotenv.config();
+
+const serviceBusConnectionString =
+  env["SERVICEBUS_CONNECTION_STRING"] || "<service bus connection string not in environment>";
+
+// NOTE: this sample uses a queue but would also work a session enabled subscription.
+const queueName = env["QUEUE_NAME_WITH_SESSIONS"] || "<queue name not in environment>";
+
+const maxSessionsToProcessSimultaneously = 8;
+const sessionIdleTimeoutMs = 3 * 1000;
+const delayWhenNoSessionsAvailableMs = 5 * 1000;
+
+// this can be used control when the round-robin processing will terminate.
+const abortController = new AbortController();
+
+// called just before we start processing the first message of a session
+async function processInitialize(sessionId: string) {
+  console.log(`[${sessionId}] will start processing...`);
+}
+
+// called when we get a message for a session
+async function processMessage(msg: ReceivedMessageWithLock) {
+  console.log(`[${msg.sessionId}] received message with body ${msg.body}`);
+}
+
+// called if we get an error
+async function processError(err: Error, sessionId?: string) {
+  console.log(`[${sessionId}] had error ${err.message}`);
+}
+
+// Called if we are closing a session
+// `reason` will be:
+// * 'error' if we are closing because of an error(the error will be delivered
+//   to `processError` above)
+// * 'idle_timeout' if we `sessionIdleTimeoutMs` milliseconds pass without
+//   any messages being received (ie, session can be considered empty).
+//
+async function processClose(reason: "error" | "idle_timeout", sessionId: string) {
+  if (reason === "error") {
+    console.log(`[${sessionId}] was closed because of error`);
+  } else if (reason === "idle_timeout") {
+    console.log(`[${sessionId}] was closed - no more messages within idle timeout`);
+  }
+}
+
+// utility function to create a timer that can be refreshed
+function createRefreshableTimer(timeoutMs: number, resolve: Function): () => void {
+  let timer: any;
+
+  return () => {
+    clearTimeout(timer);
+    timer = setTimeout(() => resolve(), timeoutMs);
+  };
+}
+
+// Queries Service Bus for the next available session and processes it.
+async function receiveFromNextSession(serviceBusClient: ServiceBusClient): Promise<void> {
+  let sessionReceiver: SessionReceiver<ReceivedMessageWithLock>;
+
+  try {
+    sessionReceiver = await serviceBusClient.createSessionReceiver(queueName, "peekLock", {
+      autoRenewLockDurationInMs: sessionIdleTimeoutMs
+    });
+  } catch (err) {
+    if ((err as MessagingError).code === "SessionCannotBeLockedError") {
+      console.log(`INFO: no available sessions, sleeping for ${delayWhenNoSessionsAvailableMs}`);
+      await delay(delayWhenNoSessionsAvailableMs);
```

Comment:
I see this delay in the same vein as the delay we have in between retries. If so, then why is it that only the `SessionCannotBeLockedError` case gets to benefit from it? Why not all the errors?

Reply:
It's possible they do - this one is the only one I definitely knew I should delay for.

If there are others that make sense that wouldn't be covered by the normal retry logic let me know.

Reply:
There is no normal retry logic atleast at the moment. We don't recover broken receiver links when it comes to sessions. So in my eyes, all errors would need this delay. If so, then the only difference is whether you want to call `processError` for `SessionCannotBeLockedError` or not.

For the sake of simplicity, I would treat all the caught errors as the same i.e call `processError`, do the delay and return.

Reply:
Sounds like that's the best option then.